ReleaseProcessAutomation/Jira/JiraTask.cs

```diff
@@ -0,0 +1,39 @@
+﻿// Copyright (c) rubicon IT GmbH, www.rubicon.eu
+//
+// See the NOTICE file distributed with this work for additional information
+// regarding copyright ownership.  rubicon licenses this file to you under 
+// the Apache License, Version 2.0 (the "License"); you may not use this 
+// file except in compliance with the License.  You may obtain a copy of the 
+// License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software 
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT 
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the 
+// License for the specific language governing permissions and limitations
+// under the License.
+//
+
+using System;
+using RestSharp.Authenticators;
+
+namespace ReleaseProcessAutomation.Jira
+{
+  public abstract class JiraTask
+  {
+    protected IAuthenticator Authenticator;
+
+    protected JiraTask (string? jiraUsername, string? jiraPassword)
```

Comment:
The more I look at `JiraTask` the more I want to remove it completely :D 

Its only use case here is the creation of the authenticator, which seems wrong to begin with. The only usages for the authenticator are the creation of the `JiraRestClient`. So we are probably better off if we remove this class and feed in the `JiraRestClient` directly in the subclasses.

If we want to go next level we can also go one step further as the `JiraRestClient` is only used to create other services. We could move that logic into an interface, for example, `IJira` and have it create the services for us. This way the jira task no longer need to care about auth/rest whatever and simply create the services they need to do their jobs

Reply:
The `JiraTask `is gone now, but I'm not quite sure what you meant in the last paragraph. Could you add a more detailed description of what you mean?

Reply:
If you look at the usages of the `JiraRestClient` it is always passed into the other services:

    var jiraProjectVersionFinder = new JiraProjectVersionFinder (jiraRestClient);

This way, we always have hard dependencies on them and it is more difficult to test. What we would do is create an interfaces with implementation (the `IJira`) which contains all the services that are used for the interaction with Jira.

Instead of the code above the method would receive a `IJira` instead of the `JiraRestClient` and then access the `JiraProjectVersionFinder` using `jira.JiraProjectVersionFinder`. This way, we can use the interfaces and do not need to depend on the actual classes making testing easier while also removing the need for the method to know about a `JiraRestClient`.

Reply:
There is now an `IJira` interface with properties to be used like `_jira.VersionCreator.CreateVersion()`. The properties are however not inserted into the implementation, they are created at runtime. This is because otherwise we would start the application and would immediately be asked for the JiraCredentials. 

To insert them we could either ignore that problem, or create some sort of `Initialize` Method in all of the services. Do we want that?

Reply:
Constructors should not contain much more logic than setting their fields. If we prompt the user in the constructor we should change that. Looking at the code I think it is best to introduce a provider for the `JiraRestClient`.

The provider interface contains a single `JiraRestClient GetJiraRestClient()` method and the implementation does the prompting, creation, and probably also caches the instance for future requests. The `IJira` implementation then receives the provides and stores it and uses it in its methods to retrieve an instance of the `JiraRestClient`.

This way, we only set a field in the constructor and the prompting is done the first time an actual action is done.